pkg/kubectl/cmd/util/config_flags.go

```diff
@@ -0,0 +1,289 @@
+package util
+
+import (
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/pflag"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+
+	"k8s.io/kubernetes/pkg/kubectl/util/transport"
+)
+
+const (
+	flagClusterName      = "cluster"
+	flagAuthInfoName     = "user"
+	flagContext          = "context"
+	flagNamespace        = "namespace"
+	flagAPIServer        = "server"
+	flagInsecure         = "insecure-skip-tls-verify"
+	flagCertFile         = "client-certificate"
+	flagKeyFile          = "client-key"
+	flagCAFile           = "certificate-authority"
+	flagBearerToken      = "token"
+	flagImpersonate      = "as"
+	flagImpersonateGroup = "as-group"
+	flagUsername         = "username"
+	flagPassword         = "password"
+	flagTimeout          = "request-timeout"
+)
+
+type RESTClientGetter interface {
+	ToRESTClient() (*rest.RESTClient, error)
+	ToRESTConfig() (*rest.Config, clientcmd.ClientConfig, error)
+	ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error)
+}
+
+// ConfigFlags composes the set of values necessary
+// for obtaining a REST client config
+type ConfigFlags struct {
+	CacheDir   *string
+	KubeConfig *string
+
+	// config flags
+	ClusterName      *string
+	AuthInfoName     *string
+	Context          *string
+	Namespace        *string
+	APIServer        *string
+	Insecure         *bool
+	CertFile         *string
+	KeyFile          *string
+	CAFile           *string
+	BearerToken      *string
+	Impersonate      *string
+	ImpersonateGroup *[]string
+	Username         *string
+	Password         *string
+	Timeout          *string
+}
+
+// ToRESTClient implements RESTClientGetter
+func (f *ConfigFlags) ToRESTClient() (*rest.RESTClient, error) {
+	clientConfig, _, err := f.ToRESTConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return rest.RESTClientFor(clientConfig)
+}
+
+// ToRESTConfig implements RESTClientGetter.
+// Returns a REST client configuration based on a provided path
+// to a .kubeconfig file, loading rules, and config flag overrides.
+// Expects the AddFlags method to have been called.
+func (f *ConfigFlags) ToRESTConfig() (*rest.Config, clientcmd.ClientConfig, error) {
```

Comment:
> I ended up returning clientcmd.ClientConfig as well since the factory_client_access.go#DefaultNamespace method uses it to lookup the namespace

That'll be a different method.